src/users/users.controller.spec.ts

```diff
@@ -0,0 +1,63 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { UsersController } from './users.controller';
+import { UsersService } from './users.service';
+import * as consts from '../test.constants';
+import { CreateUserDto } from './dto/create-user.dto';
+import { UserDto } from './dto/user.dto';
+import { UpdateUserDto } from './dto/update-user.dto';
+import { DeleteUserDto } from './dto/delete-user.dto';
+
+// Test suite for the UsersController
+describe("UsersController Unit Tests", () => {
+    let usersController: UsersController;
+    let usersService: UsersService;
+    let module: TestingModule;
+
+    beforeEach(async () => {
+        module = await Test.createTestingModule({
+            controllers: [UsersController],
+            providers: [
+                {
+                    provide: UsersService,
+                    useFactory: () => ({
+                        create: jest.fn(CreateUserDto => UserDto),
+                        findById: jest.fn(number => UserDto),
+                        update: jest.fn((number, UpdateUserDto) => UserDto),
+                        remove: jest.fn((number, DeleteUserDto) => UserDto)
+                    })
+                }
+            ],
+        }).compile();
+
+        usersService = module.get<UsersService>(UsersService);
+        usersController = module.get<UsersController>(UsersController);
+    });
```

Comment:
What do you think about setting up the controller tests in [this](https://github.com/nestjs/nest/blob/master/sample/01-cats-app/src/cats/cats.controller.spec.ts) way? Using `spyOn` and calling `mockImplementation`. This would allow the assertion of return value types and properties which seems important to me.

Reply:
The first way I did it mocked the function and made sure it returned a type of "UserDTO." I updated it so that the mock function returns the test userdto constant I made, and then I check that the function in users.controller returns that same constant. While it's a different way of doing it then the way you linked, it should accomplish the same thing.

Reply:
I updated the way I did it to run like the way in the link you posted. It's still the same way I originally mocked the users service functions, but now it specifically returns one of the test userdto's I created and checks that the user controller functions return the correct value